pkg/multishare_reconciler/reconciler.go

```diff
@@ -0,0 +1,873 @@
+	instanceInfoMap := make(map[string]*v1alpha1.InstanceInfo)
+
+	for _, instance := range instances {
+		instanceURI, err := file.GenerateMultishareInstanceURI(instance)
+		if err != nil {
+			klog.Errorf("couldn't generate instanceURI: %v for instance %q", err, instance.Name)
+			continue
+		}
+		iiName := util.InstanceURIToInstanceInfoName(instanceURI)
+
+		instanceInfo, err := recon.instanceLister.Get(iiName)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				klog.Errorf("error getting instanceInfo %q from informer: %v", iiName, err)
+				continue
+			}
+			instanceInfo, err = recon.clientset.MultishareV1alpha1().InstanceInfos().Get(context.TODO(), iiName, metav1.GetOptions{})
+			if err != nil {
+				if !errors.IsNotFound(err) {
+					klog.Errorf("error getting instanceInfo %q from api server: %v", iiName, err)
+					continue
+				}
+				klog.V(6).Infof("instanceInfo object for instance %q not found in API server", instance.Name)
+				instanceInfo = nil
+			}
+			klog.V(6).Infof("instanceInfo object for instance %q not found in informer cache but found in API server", instance.Name)
+		}
+
+		instanceInfo, err = recon.reconstructInstanceInfo(iiName, instance, instanceInfo)
+		if err != nil {
+			klog.Errorf("error creating InstanceInfo %q: %v", iiName, err)
+			continue
+		}
+		instanceInfoMap[instanceURI] = instanceInfo
+
+		instanceInfo, err = recon.maybeUpdateInstanceInfoStatus(instance, instanceInfo, instanceShares)
+		if err != nil {
+			klog.Errorf("error updating instanceInfo %q: %v", iiName, err)
+			continue
+		}
+		instanceInfoMap[instanceURI] = instanceInfo
+	}
+
+	return instanceInfoMap
+}
+
+// ShareInfo objects in the returned map must be treated as read only
+func (recon *multishareReconciler) createAndUpdateShareInfos(shares []*file.Share) map[string]*v1alpha1.ShareInfo {
+	shareInfoMap := make(map[string]*v1alpha1.ShareInfo)
+
+	// create ShareInfo and InstanceInfo that are not reflected
+	for _, share := range shares {
+		shareInfoName := util.ShareToSIName(share.Name)
+		shareInfo, err := recon.shareLister.Get(shareInfoName)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				klog.Errorf("error getting shareInfo %q from informer: %v", shareInfoName, err)
+				continue
+			}
+			shareInfo, err = recon.clientset.MultishareV1alpha1().ShareInfos().Get(context.TODO(), shareInfoName, metav1.GetOptions{})
+			if err != nil {
+				if !errors.IsNotFound(err) {
+					klog.Errorf("error getting shareInfo %q from api server: %v", shareInfoName, err)
+					continue
+				}
+				// shareInfo does not exist for share
+				klog.V(6).Infof("shareInfo object for share %q not found in API server", shareInfoName)
+				shareInfo = nil
+			}
+			// shareInfo exist in api server but not cache
+			klog.V(6).Infof("shareInfo object for share %q not found in informer cache but found in api server", shareInfoName)
```

Comment:
My thought process is logs which help provide insights in prod workflow, and which would not inundate the log file, can be V(4). I am ok with doing a separate log review, but it would be a big task in itself, so lets add logs at V(4) now which we think are useful anyway